pylabnet/launchers/launch_control.py

```diff
@@ -826,54 +894,42 @@ def main_staticproxy():
 def run(log_controller):
     """ Runs the launch controller once a Controller is instantiated"""
 
-    # Instantiate GUI
-    if not log_controller.proxy:
-        log_controller.start_logger()
-    log_controller.initialize_gui()
-    log_controller.start_gui_server()
-
-    # Start logging os in master mode
-    if log_controller.master:
-        log_controller.start_stop_logging(master_log=True)
-
-    # Standard operation
-    while not log_controller.main_window.stop_button.isChecked():
-
-        # For proxy launch controller, just check main GUI for updates
-        if log_controller.proxy:
-            log_controller.update_proxy()
-        else:
-
-            # Handle external configuration via GUI server
-            log_controller.main_window.configure_widgets()
-            log_controller.main_window.update_widgets()
-
-            # New terminal input
-            if sys.stdout.getvalue() != '':
-
-                # Check for disconnection events
-                log_controller.check_disconnection()
+    # Refresh thread
+    update_thread = QtCore.QThread()
+    
+    if log_controller.proxy:
 
-                # Handle new connections
-                log_controller.update_connection()
+        # Set up GUI
+        log_controller.initialize_gui()
+        log_controller.start_gui_server()
 
-                # Update terminal
-                log_controller.update_terminal()
+        # Set up update thread
+        updater = ProxyUpdater(log_controller)
+        updater.update_signal.connect(log_controller.update_proxy)
 
-            # Handle disconnection events
-            if log_controller.disconnection:
-                log_controller.disconnect()
+    else:
+        # Redirect sys.stdout to queue
+        queue = Queue()
+        sys.stdout = WriteStream(queue)
+        
+        # Instantiate GUI
+        log_controller.start_logger()
+        log_controller.initialize_gui()
+        log_controller.start_gui_server()
 
-        # Update display
-        log_controller.main_window.force_update()
+        # Start logging os in master mode
+        if log_controller.master:
+            log_controller.start_stop_logging(master_log=True)
 
-    # Exit, close servers if necessary
-    if log_controller.proxy:
-        pass
-    else:
-        log_controller.kill_servers()
+        # Start thread to listen for updates
+        updater = UpdateReceiver(queue)
+        updater.update_signal.connect(log_controller.update)
```

Comment:
@mkbhaskar , do I understand it correctly that the `UpdateReceiver` is launched in a new thread, which runs continuously and can communicate with the parent thread by emitting a signal, which you can then connect to a function defined in the parent thread?

Reply:
yeah, the reason being that external threads themselves are not allowed to directly call gui functions. So instead you need to use signals/slots to communicate between threads.

the alternative of course is to do all the processing in the main gui thread, but then you make a tradeoff between performance of the rapidly refreshing task and the gui responsiveness. this way, the gui can run as desired in its `app.exec_()` loop and you can independently control the rate at which the update process occurs

Reply:
Ok, I think your implementation makes sense. Let's try it together with our new master-logger infrastructure and monitor performance.